dashboard/lib/services/lti.rb

```diff
@@ -4,231 +4,236 @@
 require 'authentication_option'
 require 'sections/section'
 
-class Services::Lti
-  def self.initialize_lti_user(id_token)
-    user_type = Policies::Lti.get_account_type(id_token[Policies::Lti::LTI_ROLES_KEY])
-    user = User.new
-    user.provider = User::PROVIDER_MIGRATED
-    user.user_type = user_type
-    if user_type == User::TYPE_TEACHER
-      user.age = '21+'
-      user.name = get_claim_from_list(id_token, Policies::Lti::TEACHER_NAME_KEYS)
-      user.lti_roster_sync_enabled = true
-    else
-      user.name = get_claim_from_list(id_token, Policies::Lti::STUDENT_NAME_KEYS)
-      user.family_name = get_claim(id_token, :family_name)
+module Services
+  module Lti
+    def self.initialize_lti_user(id_token)
+      user_type = Policies::Lti.get_account_type(id_token[Policies::Lti::LTI_ROLES_KEY])
+      user = ::User.new
+      user.provider = ::User::PROVIDER_MIGRATED
+      user.user_type = user_type
+      if user_type == ::User::TYPE_TEACHER
+        user.age = '21+'
+        user.name = get_claim_from_list(id_token, Policies::Lti::TEACHER_NAME_KEYS)
+        user.lti_roster_sync_enabled = true
+      else
+        user.name = get_claim_from_list(id_token, Policies::Lti::STUDENT_NAME_KEYS)
+        user.family_name = get_claim(id_token, :family_name)
+      end
+      ao = AuthenticationOption.new(
+        authentication_id: Services::Lti::AuthIdGenerator.new(id_token).call,
+        credential_type: AuthenticationOption::LTI_V1,
+        email: get_claim(id_token, :email),
+      )
+      user.authentication_options = [ao]
+      # TODO As final step of the LTI user creation, create LtiUserIdentity for the new user. https://codedotorg.atlassian.net/browse/P20-788
+      user
     end
-    ao = AuthenticationOption.new(
-      authentication_id: Policies::Lti.generate_auth_id(id_token),
-      credential_type: AuthenticationOption::LTI_V1,
-      email: get_claim(id_token, :email),
-    )
-    user.authentication_options = [ao]
-    user
-  end
 
-  def self.create_lti_user_identity(user)
-    auth_option = user.authentication_options.find(&:lti?)
-    issuer, client_id, subject = auth_option.authentication_id.split('|')
-    lti_integration = Queries::Lti.get_lti_integration(issuer, client_id)
-    LtiUserIdentity.create(user: user, subject: subject, lti_integration: lti_integration)
-  end
+    def self.create_lti_user_identity(user)
+      auth_option = user.authentication_options.find(&:lti?)
+      issuer, client_id, subject = auth_option.authentication_id.split('|')
+      lti_integration = Queries::Lti.get_lti_integration(issuer, client_id)
+      identity = LtiUserIdentity.new(user: user, subject: subject, lti_integration: lti_integration)
+      identity.save!
```

Comment:
Why split this into two `.new` and `.save!`? For this case, wouldn't just using `.create!` be sufficient, since we aren't modifying the newed up `identity` object here?

Reply:
Both `create!` and `save!` will raise a `ActiveRecord::RecordInvalid` if it's missing required values.

Reply:
I did not touch this part of the code, I only added an indent.

Reply:
The old code on the left on line 34 calls `.create`, which instantiates the new LtiUserIdentity and saves it in a single command. In this new code, on line 36 and 37, it has been changed to use `.new` and `.save!`. I was asking why we moved away from using `.create`. If we want to raise an exception for a missing value, we can just add the `!` to the end of `.create`.